common/lib/common-utils/src/assert.ts

```diff
@@ -4,18 +4,64 @@
  */
 
 /**
- * A browser friendly version of the node assert library. Use this instead of the 'assert' package, which has a big
- * impact on bundle sizes.
+ * A browser friendly assert library.
+ * Use this instead of the 'assert' package, which has a big impact on bundle sizes.
  * @param condition - The condition that should be true, if the condition is false an error will be thrown.
  * @param message - The message to include in the error when the condition does not hold.
- * A number should not be specificed manually: use a string.
+ * A number should not be specified manually: use a string.
  * Before a release, policy-check should be run, which will convert any asserts still using strings to
  * use numbered error codes instead.
  */
 export function assert(condition: boolean, message: string | number): asserts condition {
 	if (!condition) {
-		throw new Error(
-			typeof message === "number" ? `0x${message.toString(16).padStart(3, "0")}` : message,
-		);
+		// Note that this can not simply delegate to Fail below,
+		// since that would mean calling `fail` with a message thats not a constant, which is forbidden.
+		throw new Error(formatMessage(message));
 	}
 }
+
+/**
+ * Fails an assertion.
+ * `fail(message)` is the same as `assert(false, message)`.
+ * See {@link assert}.
+ *
+ * @remarks
+ * This can be useful for raising an error when a particular code-path is reached, for example:
+ *
+ * ```typescript
+ *	switch (x) {
+ *		case "x": {
+ *			return false;
+ *		}
+ *		case "y": {
+ *			return true;
+ *		}
+ *		default: {
```

Comment:
FWIW, `unreachableCase` is usually the better choice for non-exhaustive switch statements. A possible alternative case to highlight is when the cost of reporting the error significantly outweighs that of the happy case, which favors a structure like

```typescript
if (myFastCheckForErrorCondition) {
    logger.sendErrorEvent(doHeavyComputation());
    fail("invariant violated");
}
```

to inline asserting

Reply:
Yeah, I noticed the same thing. It seems like if we logged the `never` parameter in a uniform way at all call sites we could eliminate the templated string pattern and get a small bundle size win.